File: lua_client/dh.lua
require("bint")

Dh = {
	q = bint.new(0),--素数q
    a = bint.new(0),--q的原根a
    x = bint.new(0),--私钥
    Y1 = bint.new(0),--自己公钥
    Y2 = bint.new(0)--对方公钥
}

function Dh:new(q, a)
	o = {}
    setmetatable(o, self)
    self.__index = self
    self.q = bint.new(q)
    self.a = bint.new(a)
    self.x = bint.new(0)
    self.Y1 = bint.new(0)
    self.Y2 = bint.new(0)
	self:generatePrik()
	self:generatePubk()
    return o
end

function Dh:ExchangePubk(key)
	self.Y2 = bint.new(key)
end

function Dh:ShareKey()
	local c = bint.new(0)
	local d = bint.new(0)
    local k = self.Y2 ^ self.x
	bint.div(k, self.q, c, d)
	k = d
    return bint.mt.tostring(k)
end

function Dh:PubKey()
	return bint.mt.tostring(self.Y1)
end

function Dh:generatePrik()
	self.x = bint.new(math.floor(math.random(1, 10000000))) % self.q
	self.x = self.x + 1
end

function Dh:generatePubk()
    self.Y1 = self.a ^ self.x
	self.Y1 = self.Y1 % self.q
end

DH = Dh:new(97, 5)
print("dh key is :", DH:PubKey())

File: js_client/dh.js
//dh.js
const BigMath ={
    pow(num, pow){
        let total;
        for(let i = 0; i < pow; i++){
            if(!total) total = num;
            else total = total * num;
        }
        return total;
    }
} 

function Dh(q=97n, a=5n){
    this.q = q;//素数q
    this.a = a;//q的原根a
    this.x = 0n;//私钥
    this.Y1 = 0n;//自己公钥
    this.Y2 = 0n;//对方公钥
    this.generatePrik();
    this.generatePubk();
}

Dh.prototype.ExchangePubk =function(key){
    this.Y2 = BigInt(key);
}

Dh.prototype.ShareKey =function(){
    k = BigMath.pow(this.Y2, this.x);
    k = k % this.q;
    return parseInt(BigInt.asUintN(64, k));
}

Dh.prototype.PubKey =function(){
    return parseInt(BigInt.asUintN(64, this.Y1));
}

Dh.prototype.generatePrik = function(){
    this.x = BigInt(Math.floor(Math.random()*10000000))%this.q;
    this.x = this.x + 1n;
}

Dh.prototype.generatePubk = function(){
    this.Y1 = BigMath.pow(this.a, this.x);
    this.Y1 = this.Y1 % this.q;
}

var key = new Dh();
module.exports=Dh;
module.exports.key =key;